### ISSUE TITLE
Performance Issue: Viper calls on critical paths

### ISSUE THREAD

**Issue by bcbrock** [hyperledger-archives/fabric#647]
Although it is very early to focus on performance, I noticed this today and thought it was worth documenting.

I was doing stress tests on a system with a single peer, using NOOPS consensus, invoking chaincode_example02 using the REST API from multiple clients. I used the GO profiler to get a baseline profile and was surprised to see that 15% of the run time was being consumed by the spf13/viper package calls used to get flag/parameter values. The profile indicated that fully 9% of the runtime was spent in the viper.GetBool call of this routine:

```
//ExecuteTransaction executes transactions decides to do execute in dev or prod mode
func (p *PeerImpl) ExecuteTransaction(transaction *pb.Transaction) *pb.Response {
        peerAddress := getValidatorStreamAddress()
        var response *pb.Response
        if viper.GetBool("peer.validator.enabled") { // send gRPC request to yourself
                response = sendTransactionsToThisPeer(peerAddress, transaction)

        } else {
                response = p.SendTransactionsToPeer(peerAddress, transaction)
        }

        return response
}
```

Although these viper calls will become less and less important for more complex workloads, I don't think it's unreasonable to try to avoid recomputing complex constants on critical code paths.

What is most interesting is why these calls take so long. Besides the fact that viper has to hierarchically parse a string like "peer.validator.enabled" on each call, it turns out that every time you call viper for a flag value, viper is fully formatting beaucoup logging/debug messages, _and then sending them to /dev/null_!

If in the future you want to check whether or how much option processing is occurring, add

```
jww "github.com/spf13/jwalterweatherman"
```

to the imports, and add the line

```
jww.SetStdoutThreshold(jww.LevelTrace)
```

at the beginning of a function you want to study.

**Comment by binhn** [comment]
@bcbrock good info -- i remember in Java we avoided allocating strings in the critical loops.

**Comment by corecode** [comment]
I can confirm this. How should we restructure these global settings?  Right now we're using viper as a (singleton) implicit global setting.

Should we have a `openchain/config` module, which caches important settings from viper?

**Comment by bcbrock** [comment]
@corecode, However you all decide to do this will be very rewarding in terms of performance improvement. The ideal solution will need to extend beyond Viper though. For example we see that apart from Viper, 10+ percent of runtime in a simple benchmark is spent in a system call that determines the hostname when `peer.addressAutoDetect` is true. So the entire result of `peer.GetLocalAddress()` should probably be cached.

**Comment by christo4ferris** [comment]
@binhn @bcbrock @corecode this seems to have been at least partially addressed. Certainly, the specific cited example is no longer making use of viper. Maybe this can be closed?

**Comment by bcbrock** [comment]
@christo4ferris People are still adding/modifying code that makes explicit use of variables that have been cached. It doesn't seem to be a performance issue anymore, but it is inconsistent and could lead back to one again in the future. It occurred to me to hack Viper to make it panic if people used cached variables explicitly (after they were cached) as a way to catch these. We could also make it a test in CI - it would be the simple count of matches of a particular grep (there are 2 instances of this that are "legal").

**Comment by binhn** [comment]
@bcbrock I think a test as you mentioned would be preferred so that we wouldn't have to hack Viper. If hacking Viper, then I would rather hack it to cache.  Either way, I think we should close this one and open a different issue for creating the test. @bcbrock Could you take that.

**Comment by bcbrock** [comment]
Fixed; Replaced by #1665
